### PR TITLE
Add miscellaneous unit tests

### PR DESCRIPTION
**Pull request by ryantwolf**
## Description

Adds miscellaneous unit tests to test

- `nemo_curator/services`
- `nemo_curator/nemo_run`
- `nemo_curator/utils/distributed_utils.py`
- `nemo_curator/utils/file_utils.py`
- reactivated aegis tests in `tests/test_classifiers.py`

## Usage

N/A
## Checklist

- [x] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [x] New or Existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: tests/utils/test_config_utils.py [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+        mock_filter_instance = MagicMock()
+        mock_filter_class.return_value = mock_filter_instance
+        mock_filter_instance.name = "test_filter"
+        mock_filter_instance._name = "test_filter"
+
+        # Create filter config without params
+        filter_config = {
+            "name": "nemo_curator.filters.test_filter",
+            "input_field": "text",
+        }
```
Comment (sarahyurick) [inline tests/utils/test_config_utils.py:86]: Can we also test the case when "params" is None?

File: tests/utils/test_config_utils.py [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+            "input_field": "text",
+            "filters": [
+                {
+                    "name": "nemo_curator.filters.filter1",
+                    "params": {"param1": "value1"},
+                },
+                {
+                    "name": "nemo_curator.filters.filter2",
+                    "params": {"param2": "value2"},
+                    "input_field": "custom_field",
```
Comment (sarahyurick) [inline tests/utils/test_config_utils.py:201]: Can we also test `"input_field": None`?

File: tests/utils/test_download_utils.py [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+        ]
+
+        # The function should raise ValueError when encountering non-standard format
+        with pytest.raises(ValueError) as excinfo:
+            get_main_warc_paths(mock_index, "2021-04", "2021-25")
+
+        # Check error message
+        assert "invalid literal for int()" in str(excinfo.value)
+
+    def test_filters_by_year(self):
```
Comment (sarahyurick) [inline tests/utils/test_download_utils.py:103]: Want to make sure this is testing both conditions of `if beg_year < 2013 or end_year < 2013`?

File: tests/utils/test_distributed_utils.py [hunk 4]
```diff
@@ -0,0 +1,851 @@
...
+    yield temp_dir
+    shutil.rmtree(temp_dir)
+
+
+class TestClientFunctions:
+    @pytest.mark.gpu
+    @patch("nemo_curator.utils.distributed_utils.LocalCUDACluster")
+    @patch("nemo_curator.utils.distributed_utils.Client")
+    @patch("nemo_curator.utils.distributed_utils._set_torch_to_use_rmm")
+    def test_start_dask_gpu_local_cluster(
```
Comment (sarahyurick) [inline tests/utils/test_distributed_utils.py:75]: Does this cover the `if enable_spilling` block and `assert get_num_workers(client) > 0, "No workers are currently connected."` check?

File: tests/utils/test_distributed_utils.py [hunk 5]
```diff
@@ -0,0 +1,851 @@
...
+            # Verify the client was created with the right cluster
+            mock_client.assert_called_once_with(mock_cluster_instance)
+            assert client == mock_client_instance
+
+            # Verify _set_torch_to_use_rmm was called
+            mock_set_torch_rmm.assert_called_once()
+
+    @patch("nemo_curator.utils.distributed_utils.LocalCluster")
+    @patch("nemo_curator.utils.distributed_utils.Client")
+    def test_start_dask_cpu_local_cluster(self, mock_client, mock_local_cluster):
```
Comment (sarahyurick) [inline tests/utils/test_distributed_utils.py:101]: Does this check `assert get_num_workers(client) > 0, "No workers are currently connected."`?

File: tests/utils/test_distributed_utils.py [hunk 6]
```diff
@@ -0,0 +1,851 @@
...
+        # Test with scheduler_address
+        mock_client_instance = MagicMock()
+        mock_client.return_value = mock_client_instance
+
+        # Mock get_num_workers to return a positive value
+        with patch(
+            "nemo_curator.utils.distributed_utils.get_num_workers", return_value=1
+        ):
+            # Test with scheduler_address
+            client = get_client(scheduler_address="tcp://localhost:8786")
```
Comment (sarahyurick) [inline tests/utils/test_distributed_utils.py:134]: Add case when both `scheduler_address` and `scheduler_file` are specified?

Comment (sarahyurick) [reply tests/utils/test_distributed_utils.py:134]: Check `assert get_num_workers(client) > 0, "No workers are currently connected."`?

File: tests/utils/test_distributed_utils.py [hunk 7]
```diff
@@ -0,0 +1,851 @@
...
+            mock_client.assert_called_with(
+                address="tcp://localhost:8786", timeout="30s"
+            )
+
+            # Test with scheduler_file
+            client = get_client(scheduler_file="/path/to/scheduler.json")
```
Comment (sarahyurick) [inline tests/utils/test_distributed_utils.py:140]: Check `assert get_num_workers(client) > 0, "No workers are currently connected."`?

File: tests/utils/test_distributed_utils.py [hunk 8]
```diff
@@ -0,0 +1,851 @@
...
+        pd.testing.assert_frame_equal(result, df[["a", "custom_name"]])
+
+        # Test with parquet type (should not filter)
+        result = select_columns(df, ["a"], "parquet", False)
+        assert result is df
+
+    @pytest.mark.gpu
+    @patch("nemo_curator.utils.distributed_utils.cudf")
+    @patch("nemo_curator.utils.distributed_utils.pd")
+    def test_read_single_partition(self, mock_pd, mock_cudf, temp_dir):
```
Comment (sarahyurick) [inline tests/utils/test_distributed_utils.py:219]: Check `if input_meta is not None and file_type != "jsonl"` warning?

File: tests/utils/test_distributed_utils.py [hunk 9]
```diff
@@ -0,0 +1,851 @@
...
+        mock_read_blocksize.return_value = dd.from_pandas(
+            pd.DataFrame({"a": [1, 2]}), npartitions=2
+        )
+
+        # Test reading pickle
+        with patch(
+            "nemo_curator.utils.distributed_utils.dd.from_pandas",
+            return_value=dd.from_pandas(mock_read_pickle.return_value, npartitions=16),
+        ):
+            result = read_data(["file.pkl"], file_type="pickle")
```
Comment (sarahyurick) [inline tests/utils/test_distributed_utils.py:393]: Can test when the input is a string instead of a list.

File: tests/utils/test_distributed_utils.py [hunk 10]
```diff
@@ -0,0 +1,851 @@
...
+            mock_read_pickle.assert_called_once()
+
+        # Test reading jsonl with blocksize
+        result = read_data(
+            ["file1.jsonl", "file2.jsonl"],
+            file_type="jsonl",
+            blocksize="1MB",
+            files_per_partition=None,  # Explicitly set to None to avoid conflict
```
Comment (sarahyurick) [inline tests/utils/test_distributed_utils.py:401]: Add case when both of these are set.

File: tests/utils/test_distributed_utils.py [hunk 11]
```diff
@@ -0,0 +1,851 @@
...
+            with pytest.raises(RuntimeError):
+                read_data(["file1.jsonl", "file2.parquet"], file_type="jsonl")
+
+        # Test invalid file type
+        with pytest.raises(RuntimeError):
+            read_data(["file.txt"], file_type="txt")
+
+
+class TestDataWritingFunctions:
+    def test_single_partition_write_with_filename(self, temp_dir):
```
Comment (sarahyurick) [inline tests/utils/test_distributed_utils.py:425]: Should we test cuDF versus Pandas?

And `elif output_type == "bitext"`?

File: tests/utils/test_distributed_utils.py [hunk 12]
```diff
@@ -0,0 +1,851 @@
...
+        # Test with unknown output type
+        with pytest.raises(ValueError):
+            single_partition_write_with_filename(
+                df,
+                os.path.join(temp_dir, "output"),
+                output_type="unknown",
+                filename_col="file_name",
+            )
+
+    @patch("nemo_curator.utils.distributed_utils._write_to_jsonl_or_parquet")
```
Comment (sarahyurick) [inline tests/utils/test_distributed_utils.py:474]: I think there may be blocks of `_write_to_jsonl_or_parquet` that are not covered?

File: tests/utils/test_distributed_utils.py [hunk 13]
```diff
@@ -0,0 +1,851 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
```
Comment (sarahyurick) [inline tests/utils/test_distributed_utils.py:1]: Add `get_gpu_memory_info` test?

File: tests/utils/test_file_utils.py [hunk 14]
```diff
@@ -0,0 +1,820 @@
...
+class TestFileProcessingUtils:
+    """Tests for file processing utility functions."""
+
+    def test_get_remaining_files(self, temp_dir):
+        """Test get_remaining_files function."""
+        # Create input and output directories
+        input_dir = os.path.join(temp_dir, "input")
+        output_dir = os.path.join(temp_dir, "output")
+        os.makedirs(input_dir)
+        os.makedirs(output_dir)
```
Comment (sarahyurick) [inline tests/utils/test_file_utils.py:249]: Can we include the case when the output directory does not already exist?

File: tests/utils/test_file_utils.py [hunk 15]
```diff
@@ -0,0 +1,820 @@
...
+            parse_str_of_num_bytes("invalid")
+
+        with pytest.raises(ValueError):
+            parse_str_of_num_bytes("1x")
+
+
+class TestJSONLProcessing:
+    """Tests for JSONL processing utility functions."""
+
+    def test_save_jsonl(self, temp_dir):
```
Comment (sarahyurick) [inline tests/utils/test_file_utils.py:734]: When would `except Exception as exception` be raised for the `_save_jsonl` function?

**Review by sarahyurick — COMMENTED** [review]
Very extensive tests, thanks!

File: tests/test_classifiers.py [hunk 16]
```diff
@@ -86,9 +88,6 @@ def test_quality_classifier(gpu_client):
     assert result_pred.equals(expected_pred)
 
 
-@pytest.mark.skip(
-    reason="Aegis needs to be downloaded and cached to our gpuCI runner to enable this"
-)
```
[the remark this answers is not in the document]
Comment (ryantwolf) [reply tests/test_classifiers.py:91]: yep for sure. I think this is a potential problem with all of our classifiers.

File: tests/utils/test_download_utils.py (hunk 3 above)
Comment (ryantwolf) [reply tests/utils/test_download_utils.py:103]: Added one more test for when end year is less than 2013

File: tests/utils/test_distributed_utils.py (hunk 4 above)
Comment (ryantwolf) [reply tests/utils/test_distributed_utils.py:75]: Yes it does those checks.

File: tests/utils/test_distributed_utils.py (hunk 5 above)
Comment (ryantwolf) [reply tests/utils/test_distributed_utils.py:101]: Yes

File: tests/utils/test_distributed_utils.py (hunk 7 above)
Comment (ryantwolf) [reply tests/utils/test_distributed_utils.py:140]: This assert statement (or the others like it) is passing in all these tests.

File: tests/utils/test_distributed_utils.py (hunk 11 above)
Comment (ryantwolf) [reply tests/utils/test_distributed_utils.py:425]: I think the cudf vs pandas tests are in `test_io.py`.  The bitext path is tested in `test_bitext_distributed_utils.py`

**Review by sarahyurick — APPROVED** [review]
Looks like everything was addressed. Thanks!
